### ISSUE TITLE
Extend generic static bug

### ISSUE THREAD

**Issue by mushishi78** [Microsoft/TypeScript#17829]
**TypeScript Version:**  2.4.0

**Code**

```ts
function base<T>() {
    class Base {
        static prop: T;
    }
    return Base;
}

class Gen<T> extends base<T>() {}
class Spec extends Gen<string> {}

<string>Spec.prop;
```

**Expected behavior:** Either the static property `prop` is of type string or I'm given an error to prevent this happening earlier with clear error message.

**Actual behavior:** It claims that `prop` is of type `T`.

**Comment by ahejlsberg** [comment]
The issue here is that type parameters introduced by a class declaration should not be in scope _within_ the expression part of the base class heritage clause. However, when a type argument list _follows_ the expression part, they should be in scope there. So, this should be an error

```ts
class Gen<T> extends base<T>() {}  // Error, T not in scope in base class expression
```

**Comment by markboyall** [comment]
Then... how would you implement the Gen class above?

**Comment by ahejlsberg** [comment]
@markboyall Yes, you'd need to wrap the derived class in a function as well such that you can pass a type parameter that differs _per invocation_. For example:

```ts
function base<T>() {
    class Base {
        static prop: T;
    }
    return Base;
}

function derived<T>() {
    class Derived extends base<T>() {
        static anotherProp: T;
    }
    return Derived;
}

class Spec extends derived<string>() {}

Spec.prop;  // string
Spec.anotherProp;  // string
```

The lifetime of the statics above is one per invocation of their containing function. Effectively, a class declaration manufactures a class constructor function each time it is executed (which is only once for a global class, but can be many times for a class nested in a function).
